src/pytest_bdd/generation.py

```diff
@@ -100,18 +103,34 @@ def print_missing_code(scenarios: list[ScenarioTemplate], steps: list[Step]) ->
         tw.line()
         if step.scenario is not None:
             tw.line(
-                """Step {step} is not defined in the scenario "{step.scenario.name}" in the feature"""
-                """ "{step.scenario.feature.name}" in the file"""
-                """ {step.scenario.feature.filename}:{step.line_number}""".format(step=step),
+                (
+                    f'Step {step} is not defined in the scenario "{step.scenario.name}" '
+                    f'in the feature "{step.scenario.feature.name}" in the file '
+                    f"{step.scenario.feature.filename}:{step.line_number}"
+                ),
                 red=True,
             )
         elif step.background is not None:
-            tw.line(
-                """Step {step} is not defined in the background of the feature"""
-                """ "{step.background.feature.name}" in the file"""
-                """ {step.background.feature.filename}:{step.line_number}""".format(step=step),
-                red=True,
-            )
+            feature_or_rule = step.background.parent
+            if isinstance(feature_or_rule, Feature):
+                tw.line(
+                    (
+                        f"Step {step} is not defined in the background of the feature "
+                        f'"{feature_or_rule.name}" in the file '
+                        f"{feature_or_rule.filename}:{step.line_number}"
+                    ),
+                    red=True,
+                )
+            elif isinstance(feature_or_rule, Rule):
+                _feature: Feature = feature_or_rule.feature
+                tw.line(
+                    (
+                        f"Step {step} is not defined in the background of the rule "
+                        f'"{feature_or_rule.name}" in the file '
+                        f"{_feature.filename}:{step.line_number}"
+                    ),
+                    red=True,
+                )
```

Comment:
is the only thing different the message? If so, let's have the message be conditionally defined from the `isinstance` check, rather than this huge blob.
something like:
```python
if isinstance(parent, Rule):
    parent_type = "rule"
    filename = parent
elif isinstance(parent, Feature):
    parent_type = "feature"
    filename = parent._filename
else:
    # raise NotImplementedError?
message =  (
    f"Step {step} is not defined in the background of the {parent_type} "
    f'"{parent.name}" in the file {filename}:{step.line_number}'
)
tw.line(message, ...)
```